pkg/controller/podautoscaler/replica_calculator.go

```diff
@@ -0,0 +1,246 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podautoscaler
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	unversionedcore "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/unversioned"
+	metricsclient "k8s.io/kubernetes/pkg/controller/podautoscaler/metrics"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+type ReplicaCalculator struct {
+	metricsClient metricsclient.MetricsClient
+	podsGetter    unversionedcore.PodsGetter
+}
+
+func NewReplicaCalculator(metricsClient metricsclient.MetricsClient, podsGetter unversionedcore.PodsGetter) *ReplicaCalculator {
+	return &ReplicaCalculator{
+		metricsClient: metricsClient,
+		podsGetter:    podsGetter,
+	}
+}
+
+// GetResourceReplicas calculates the desired replica count based on a target resource utilization percentage
+// of the given resource for pods matching the given selector in the given namespace, and the current replica count
+func (c *ReplicaCalculator) GetResourceReplicas(currentReplicas int32, targetUtilization int32, resource api.ResourceName, namespace string, selector labels.Selector) (int32, int32, time.Time, error) {
+	metrics, timestamp, err := c.metricsClient.GetResourceMetric(resource, namespace, selector)
+	if err != nil {
+		return 0, 0, time.Time{}, fmt.Errorf("unable to get metrics for resource %s: %v", resource, err)
+	}
+
+	podList, err := c.podsGetter.Pods(namespace).List(api.ListOptions{LabelSelector: selector})
+	if err != nil {
+		return 0, 0, time.Time{}, fmt.Errorf("unable to get pods while calculating replica count: %v", err)
+	}
+
+	if len(podList.Items) == 0 {
+		return 0, 0, time.Time{}, fmt.Errorf("no pods returned by selector while calculating replica count")
+	}
+
+	requests := make(map[string]int64, len(podList.Items))
+	readyPodCount := 0
+	unreadyPods := sets.NewString()
+	missingPods := sets.NewString()
+
+	for _, pod := range podList.Items {
+		podSum := int64(0)
+		for _, container := range pod.Spec.Containers {
+			if containerRequest, ok := container.Resources.Requests[resource]; ok {
+				podSum += containerRequest.MilliValue()
+			} else {
+				return 0, 0, time.Time{}, fmt.Errorf("missing request for %s on container %s in pod %s/%s", resource, container.Name, namespace, pod.Name)
+			}
+		}
+
+		requests[pod.Name] = podSum / int64(len(pod.Spec.Containers))
```

Comment:
good catch.  Weird that the unit tests didn't catch that.  I'll make sure there's a test which exercises that.